externallib/DPA/DP.py

```diff
@@ -0,0 +1,152 @@
+"""
+Module implementing the "Fast Search and Find of Density Peaks" clustering algorithm of Rodriguez and Laio (2014).
+"""
+
+import numpy as np
+import scipy as sp
+from scipy import stats
+from sklearn.neighbors import NearestNeighbors
+
+
+class Density_Peaks_clustering:
+    """
+    Clustering by fast search and find of density peaks, Rodriguez and Laio (2014).
+    """
+    def __init__(self, distances=None, indices=None, dens_type="eps", dc=None, percent=2.0):
+
+        """
+
+        Parameters
+        ----------
+        distances: Matrix of distances between data points and their neighbours of dimension Nxn_neigh_comp where N is
+        the number of data points and n_neigh_comp is the number of neighbours to consider.
+        indices: The indices in the data matrix of the neighours for all of the data points.
+        dens_type: The type of density to compute (can be exponential)
+        dc: The cutoff distance beyond which data points don't contribute to the local density computation of another
+        datapoint.
+        percent: Criterion for choosing dc, int from 0-100, typically chosen such that the average number of neighbours is 1-2% of the
+        total number of points in the dataset.
+        """
+
+        self.distances = distances
+        self.indices = indices
+        self.dens_type = dens_type
+        self.dc = dc
+        self.percent = percent
+        self.dens = None  # numpy array of the densities of the data points
```

Comment:
One conflict seems to be that the **clustering.py** script operates on the log density whereas Density_Peaks_Clustering works with the (unlogged) density.

Reply:
If there is such a transform it would be good to understand where the best place to apply it might be